// @ant-design-vue/nuxt/dist/module.mjs
import { addComponent, addImportsSources, useNuxt, defineNuxtModule, createResolver } from '@nuxt/kit';
import * as AllIcons from '@ant-design/icons-vue';

const allComponents = [
  "App",
  "Affix",
  "Alert",
  "Anchor",
  "AnchorLink",
  "AutoComplete",
  "AutoCompleteOptGroup",
  "AutoCompleteOption",
  "Avatar",
  "AvatarGroup",
  "BackTop",
  "Badge",
  "BadgeRibbon",
  "Breadcrumb",
  "BreadcrumbItem",
  "BreadcrumbSeparator",
  "Button",
  "ButtonGroup",
  "Calendar",
  "Card",
  "CardGrid",
  "CardMeta",
  "Carousel",
  "Cascader",
  "CheckableTag",
  "Checkbox",
  "CheckboxGroup",
  "Col",
  "Collapse",
  "CollapsePanel",
  "Comment",
  "ConfigProvider",
  ["Compact", "SpaceCompact"],
  "DatePicker",
  "Descriptions",
  "DescriptionsItem",
  "DirectoryTree",
  "Divider",
  "Drawer",
  "Dropdown",
  "DropdownButton",
  "Empty",
  "FloatButton",
  "Flex",
  "FloatButtonGroup",
  "Form",
  "FormItem",
  "FormItemRest",
  "Grid",
  "Image",
  "ImagePreviewGroup",
  "Input",
  "InputGroup",
  "InputNumber",
  "InputPassword",
  "InputSearch",
  "Layout",
  "LayoutContent",
  "LayoutFooter",
  "LayoutHeader",
  "LayoutSider",
  "List",
  "ListItem",
  "ListItemMeta",
  "LocaleProvider",
  "Mentions",
  "MentionsOption",
  "Menu",
  "MenuDivider",
  "MenuItem",
  "MenuItemGroup",
  "Modal",
  "MonthPicker",
  "PageHeader",
  "Pagination",
  "Popconfirm",
  "Popover",
  "Progress",
  ["QRCode", "Qrcode"],
  "QuarterPicker",
  "Radio",
  "RadioButton",
  "RadioGroup",
  "RangePicker",
  "Rate",
  "Result",
  "Row",
  "Segmented",
  "Select",
  "SelectOptGroup",
  "SelectOption",
  "Skeleton",
  "SkeletonAvatar",
  "SkeletonButton",
  "SkeletonImage",
  "SkeletonInput",
  "SkeletonTitle",
  "Slider",
  "Space",
  "Spin",
  "Statistic",
  "StatisticCountdown",
  "Step",
  "Steps",
  "SubMenu",
  "Switch",
  "StyleProvider",
  "TabPane",
  "Table",
  "TableColumn",
  "TableColumnGroup",
  "TableSummary",
  "TableSummaryCell",
  "TableSummaryRow",
  "Tabs",
  "Tag",
  "Textarea",
  "TimePicker",
  "TimeRangePicker",
  "Timeline",
  "TimelineItem",
  "Tooltip",
  "Tour",
  "Transfer",
  "Tree",
  "TreeNode",
  "TreeSelect",
  "TreeSelectNode",
  "Typography",
  "TypographyLink",
  "TypographyParagraph",
  "TypographyText",
  "TypographyTitle",
  "Upload",
  "UploadDragger",
  "Watermark",
  "WeekPicker"
];
const allImports = ["message", "notification", "Modal", "App"];

const libraryName = "ant-design-vue";
const prefix = "A";
const allIcons = Object.keys(AllIcons).filter((v) => /.*(Outlined|Filled|TwoTone)$/.test(v));
const defaults = {
  components: allComponents,
  icons: allIcons,
  imports: allImports,
  extractStyle: false
};

const resolveComponents = (config, filePath) => {
  const { components, icons } = config;
  const allComponents = components === false ? [] : components;
  allComponents.forEach((component) => {
    if (typeof component === "string") {
      addComponent({
        export: component,
        name: prefix + component,
        filePath
      });
    } else if (Array.isArray(component)) {
      addComponent({
        export: component[0],
        name: prefix + component[1],
        filePath
      });
    }
  });
  if (icons === false) {
    return;
  }
  icons.forEach((icon) => {
    if (typeof icon === "string") {
      addComponent({
        export: icon,
        name: icon,
        filePath
      });
    } else if (Array.isArray(icon)) {
      addComponent({
        export: icon[0],
        name: icon[1],
        filePath
      });
    }
  });
};

const resolveImports = (config, filePath) => {
  const { imports } = config;
  const allImports = imports ? imports : [];
  addImportsSources({
    from: filePath,
    imports: [...allImports]
  });
};

function resolveOptions() {
  const nuxt = useNuxt();
  nuxt.options.build.transpile.push(libraryName);
  if (nuxt.options.builder === "@nuxt/vite-builder") {
    nuxt.options.vite.plugins ??= [];
    nuxt.options.vite.plugins.push({
      name: "antd-dayjs",
      config() {
        return {
          resolve: {
            alias: [
              {
                find: /^(ant-design-vue)(?!\/(es|dist))/,
                replacement: "ant-design-vue/es"
              },
              {
                find: /^ant-design-vue\/dist$/,
                replacement: "ant-design-vue/dist"
              },
              {
                find: /^ant-design-vue\/es$/,
                replacement: "ant-design-vue/es"
              },
              {
                find: /^@ant-design\/icons-vue$/,
                replacement: "@ant-design/icons-vue"
              },
              {
                find: /^dayjs\/plugin\/(.*)$/,
                replacement: "dayjs/esm/plugin/$1"
              },
              {
                find: /^dayjs\/locale/,
                replacement: "dayjs/esm/locale"
              },
              {
                find: /^dayjs$/,
                replacement: "dayjs/esm"
              }
            ]
          }
        };
      }
    });
  }
}

const module = defineNuxtModule({
  meta: {
    name: libraryName,
    configKey: "antd"
  },
  // Default configuration options of the Nuxt module
  defaults,
  setup(_options, nuxt) {
    const options = _options;
    resolveOptions();
    const antdRuntimePath = createResolver(import.meta.url).resolve("./runtime/antd");
    nuxt.options.imports.autoImport !== false && resolveImports(options, antdRuntimePath);
    nuxt.options.components !== false && resolveComponents(options, antdRuntimePath);
    if (options.extractStyle) {
      extractStyle();
    }
  }
});
function extractStyle() {
  const nuxt = useNuxt();
  if (!nuxt.options.ssr) {
    return;
  }
  if (nuxt.options.dev === false && nuxt.options.nitro.static) {
    nuxt.options.nitro.replace ??= {};
    nuxt.options.nitro.replace["process.env.NODE_ENV"] = "'production'";
  }
  const resolver = createResolver(import.meta.url);
  addComponent({
    name: "AExtractStyle",
    filePath: resolver.resolve("./runtime/components/AExtractStyle.vue")
  });
}

export { module as default };
